zk-gadget/src/mock.rs

```diff
@@ -0,0 +1,399 @@
+            AccountId32::new([3u8; 32]),
+            AccountId32::new([4u8; 32]),
+            AccountId32::new([5u8; 32]),
+            AccountId32::new([6u8; 32]),
+            AccountId32::new([7u8; 32]),
+            AccountId32::new([8u8; 32]),
+        ];
+        validators.contains(&address)
+    }
+
+    fn report_offence(_offence_report: ReportValidatorOffence<AccountId32>) -> DispatchResult {
+        Ok(())
+    }
+
+    fn get_validator_metadata(address: AccountId32, _job_key: JobKey) -> Option<RoleTypeMetadata> {
+        let mock_err_account = AccountId32::new([100u8; 32]);
+        if address == mock_err_account {
+            None
+        } else {
+            Some(RoleTypeMetadata::ZkSaas(ZkSaasRoleMetadata::default()))
+        }
+    }
+}
+
+pub struct MockMPCHandler;
+
+impl MPCHandler<AccountId32, BlockNumber, Balance> for MockMPCHandler {
+    fn verify(_data: JobWithResult<AccountId32>) -> DispatchResult {
+        Ok(())
+    }
+
+    fn verify_validator_report(
+        _validator: AccountId32,
+        _offence: ValidatorOffenceType,
+        _signatures: Vec<Vec<u8>>,
+    ) -> DispatchResult {
+        Ok(())
+    }
+
+    fn validate_authority_key(_validator: AccountId32, _authority_key: Vec<u8>) -> DispatchResult {
+        Ok(())
+    }
+}
+
+parameter_types! {
+    pub const JobsPalletId: PalletId = PalletId(*b"py/jobss");
+}
+
+impl pallet_jobs::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type ForceOrigin = EnsureSigned<AccountId32>;
+    type Currency = Balances;
+    type JobToFee = MockJobToFeeHandler;
+    type RolesHandler = MockRolesHandler;
+    type MPCHandler = MockMPCHandler;
+    type PalletId = JobsPalletId;
+    type WeightInfo = ();
+}
+
+construct_runtime!(
+    pub enum Runtime
+    {
+        System: frame_system,
+        Timestamp: pallet_timestamp,
+        Balances: pallet_balances,
+        Jobs: pallet_jobs,
+    }
+);
+
+#[derive(Clone, Debug)]
+pub struct MockClient {
+    runtime: Runtime,
+    externalities: SharedTestExternalities,
+}
+
+#[async_trait::async_trait]
+impl Client<Block> for MockClient {
+    async fn get_next_finality_notification(&self) -> Option<FinalityNotification<Block>> {
+        self.get_latest_finality_notification().await
+    }
+
+    async fn get_latest_finality_notification(&self) -> Option<FinalityNotification<Block>> {
+        let (tx, rx) = sc_utils::mpsc::tracing_unbounded("mpsc_finality_notification", 999999);
+        // forget rx so that it doesn't get dropped
+        core::mem::forget(rx);
+        let header = self
+            .externalities
+            .lock()
+            .execute_with(|| System::finalize());
+        let summary = FinalizeSummary::<Block> {
+            finalized: vec![header.hash()],
+            header,
+            stale_heads: vec![],
+        };
+        let notification = FinalityNotification::from_summary(summary, tx.clone());
+        Some(notification)
+    }
```

Comment:
Note that the underlying AbstractGadget implementation will keep polling this in a loop since it's not actually awaiting anything. It will keep sending results in probably too fast.